source/s3_chunk_stream.c

```diff
@@ -0,0 +1,271 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include "aws/s3/private/s3_checksums.h"
+#include <aws/common/encoding.h>
+#include <aws/common/string.h>
+#include <aws/io/stream.h>
+#include <inttypes.h>
+
+AWS_STATIC_STRING_FROM_LITERAL(s_carriage_return, "\r\n");
+AWS_STATIC_STRING_FROM_LITERAL(s_empty_chunk, "0\r\n");
+AWS_STATIC_STRING_FROM_LITERAL(s_final_chunk, "\r\n0\r\n");
+AWS_STATIC_STRING_FROM_LITERAL(s_colon, ":");
+AWS_STATIC_STRING_FROM_LITERAL(s_post_trailer, "\r\n\r\n");
+
+struct aws_chunk_stream;
+
+typedef int(set_stream_fn)(struct aws_chunk_stream *parent_stream);
+
+struct aws_chunk_stream {
+    /* aws_input_stream_byte_cursor provides our actual functionality  */
+    struct aws_input_stream *current_stream;
+    struct aws_input_stream *checksum_stream;
+    struct aws_byte_buf checksum_result;
+    struct aws_byte_buf *checksum_result_output;
+    struct aws_byte_buf pre_chunk_buffer;
+    struct aws_byte_buf post_chunk_buffer;
+    const struct aws_byte_cursor *checksum_header_name;
+    int64_t length;
+    set_stream_fn *set_current_stream_fn;
+};
+
+static int s_set_null_stream(struct aws_chunk_stream *parent_stream) {
+    aws_input_stream_destroy(parent_stream->current_stream);
+    parent_stream->current_stream = NULL;
+    parent_stream->set_current_stream_fn = NULL;
+    aws_byte_buf_clean_up(&parent_stream->post_chunk_buffer);
+    return AWS_OP_SUCCESS;
+}
+
+static int s_set_post_chunk_stream(struct aws_chunk_stream *parent_stream) {
+    int64_t current_stream_length;
+    if (aws_input_stream_get_length(parent_stream->current_stream, &current_stream_length)) {
+        return AWS_OP_ERR;
+    }
+    aws_input_stream_destroy(parent_stream->current_stream);
+
+    struct aws_byte_cursor final_chunk_cursor;
+    /* should I use aws_likely here? */
+    if (current_stream_length > 0) {
+        final_chunk_cursor = aws_byte_cursor_from_string(s_final_chunk);
+    } else {
+        final_chunk_cursor = aws_byte_cursor_from_string(s_empty_chunk);
+    }
+    struct aws_byte_cursor post_trailer_cursor = aws_byte_cursor_from_string(s_post_trailer);
+    struct aws_byte_cursor colon_cursor = aws_byte_cursor_from_string(s_colon);
+
+    struct aws_byte_cursor checksum_result_cursor = aws_byte_cursor_from_buf(&parent_stream->checksum_result);
+    if (parent_stream->checksum_result_output &&
+        aws_byte_buf_init_copy_from_cursor(
+            parent_stream->checksum_result_output, aws_default_allocator(), checksum_result_cursor)) {
+        return AWS_OP_ERR;
+    }
+    if (aws_byte_buf_init(
+            &parent_stream->post_chunk_buffer,
+            aws_default_allocator(),
+            final_chunk_cursor.len + parent_stream->checksum_header_name->len + colon_cursor.len +
+                checksum_result_cursor.len + post_trailer_cursor.len)) {
+        goto error;
+    }
```

Comment:
Feel like a bug.

So, the checksum trailer header will be added to the input stream here. However, the length of the input stream will not count that. So, the length of the stream is not the same as the REAL length.

I am not sure how it will affect other things. You probably want to disable the get length of the stream as well.

Reply:
The stream needs a length, I believe it's used by http somewhere, I don't recall where exactly off the top of my head, but it's definitly used. Fortunatly this isn't a bug, as the length is calculated when creating the stream, and that value is what is returned by get length.
[https://github.com/awslabs/aws-c-s3/blob/0cb50d2b7ffb73188e97a221d75e4930067d85a7/source/s3_chunk_stream.c#L256-L259](url)

Reply:
```    /* we subtract one since aws_base64_compute_encoded_len accounts for the null terminator which won't show up in our
     * stream */
    impl->length = prechunk_stream_len + stream_length + final_chunk_len + impl->checksum_header_name->len + colon_len +
                   encoded_checksum_len + post_trailer_len - 1;```

Reply:
```
static int s_aws_input_chunk_stream_get_length(struct aws_input_stream *stream, int64_t *out_length) {
    struct aws_chunk_stream *impl = stream->impl;
    *out_length = impl->length;
    return AWS_OP_SUCCESS;
}
```

Reply:
Ah, used for the content length, right after assigning the body, note that this is also a place using a http-chunked header vs an aws-chunked header could have gone wrong.
```
int64_t stream_length = 0;
    if (aws_input_stream_get_length(input_stream, &stream_length)) {
        goto error_clean_up;
    }
    char content_length_buffer[64] = "";
    snprintf(content_length_buffer, sizeof(content_length_buffer), "%" PRIu64, (uint64_t)stream_length);
```

Reply:
okay, you are right. You know the length before the chunk sum is computed.